ouroboros-consensus-test/PLAN-statistical-testing.md

```diff
@@ -0,0 +1,159 @@
+    relatively mature, I'll replicate it for Shelley, and then eventually
+    Cardano.
+
+  - Detect CP violations so that we can count them, label them, truncate the
+    test, etc. Fail fatally if the leader schedule should have precluded CP
+    violations (eg there were no multi-leader slots).
+
+  - Empirically determine the CP violation. Use eg 100,000 iterations. (I hope
+    QuickCheck's tabulate can withstand that...)
+
+  - Implement a dedicated statistical property test. (I anticipate
+    `checkCoverage` and `coverage` and sufficiently well-behaved when used in
+    isolation.)
+
+  - Tune the confidence-level and also potentially invest time to optimize the
+    relevant mock/test infrastructure (at least check for low-hanging fruit).
+
+  - Label CPs and truncate for CPs in the Mock Praos non-statistical
+    properties. (Maybe also Shelley, even before we have dedicated Shelley
+    statistical tests?)
+
+  - Introduce semi-synchrony, ie delta greater than 1. I'm disappointed that I
+    still don't have confidence regarding how exactly to inject these delays.
+    My only insight so far that feels particularly helpful is to only delay
+    RollForward messages and let all others traveral "instantly" (this is
+    already implemented and used by the Cardano network partition). RollForward
+    is the mini-protocol that best corresponds to the "entire chain as a
+    message" used in the Ouroboros papers. However, many questions remain. EG
+    our reliable order-preserving channels can cause messages to get stuck
+    behind other messages, and -- similarly to priority inversion -- this may
+    delay a particular message beyond it's originally scheduled delay.
+
+## Miscellaneous Questions
+
+  - The above proposes focusing on the probability of a CP violation occuring
+    during some fixed number of slots. Should we instead focus on its arrival
+    rate? This would suggest choosing a "small" configuration that essentially
+    ensures a CP violation will occur within say the first 1000 slots --
+    otherwise the test may have extremely variable run-times.
+
+  - What to test in addition to the CP violation rate?
+
+      - The fraction of active slots should be f (regardless of stake
+        distribution, barring empty). Trivial to compute analytically.
+
+      - The leadership rate p_i of a node should match phi_f(its stake) (may
+        need hold the stake dist constant for this). Trivial to compute
+        analytically.
+
+      - The rate of multi-leader slots varies with f, n, and stake dist (for
+        small n) (constant stake dist again). Easy to compute analytically.
+
+      - Nodes not leading should break any ties fairly when multiple of their
+        immediate neighbors are leading simultaneously. Should be uniform, I
+        think.
+
+  - How to handle the "k+1st upstream header is one slot beyond stability
+    window" infelicity? How common is this? Easy to robustly detect? Etc.
```

Comment:
I have no idea of how common it is (depends on `k` as you said). We should definitely label it.